core-rust/state-manager/src/staging/cache.rs

```diff
@@ -66,19 +66,43 @@ use super::stage_tree::{DerivedStageKey, StageKey};
 use crate::staging::stage_tree::{Accumulator, Delta, StageTree};
 use crate::AccumulatorHash;
 use im::hashmap::HashMap as ImmutableHashMap;
-use radix_engine::ledger::{OutputValue, ReadableSubstateStore, WriteableSubstateStore};
+use radix_engine::ledger::{OutputValue, ReadableSubstateStore};
+use radix_engine::state_manager::StateDiff;
 use radix_engine::transaction::{TransactionReceipt, TransactionResult};
 use radix_engine_interface::api::types::SubstateId;
+use radix_engine_interface::crypto::{hash, Hash};
+use radix_engine_interface::data::scrypto_encode;
+use radix_engine_stores::hash_tree::put_at_next_version;
+use radix_engine_stores::hash_tree::tree_store::{
+    NodeKey, ReadableTreeStore, TreeNode, Version, WriteableTreeStore,
+};
 use sbor::rust::collections::HashMap;
 use slotmap::SecondaryMap;
 
+pub trait RootStore: ReadableSubstateStore + ReadableTreeStore {}
+impl<T: ReadableSubstateStore + ReadableTreeStore> RootStore for T {}
+
 pub struct ExecutionCache {
-    stage_tree: StageTree<TransactionReceipt, ImmutableStore>,
+    stage_tree: StageTree<ProcessedResult, ImmutableStore>,
     root_accumulator_hash: AccumulatorHash,
     accumulator_hash_to_key: HashMap<AccumulatorHash, DerivedStageKey>,
     key_to_accumulator_hash: SecondaryMap<DerivedStageKey, AccumulatorHash>,
 }
 
+pub struct ProcessedResult {
+    // the state hash is not yet stored nor exposed (WIP)
+    #[allow(dead_code)]
+    state_hash: Hash,
+    receipt: TransactionReceipt,
+    empty_state_diff: Option<StateDiff>,
```

Comment:
I see... but that's simply less convenient API for literally 100% of its consumers.

I absolutely hate the `empty_state_diff` field here, but I am lacking some rust idiom - I'd be fine if I could allocate a "statically-owned" singleton with an empty `StateDiff` instance which can be referred to from anywhere. I actually had the same problem here:
https://github.com/radixdlt/babylon-node/blob/3d5627795d1b44231d58c2722dff60fafb234433/core-rust/state-manager/src/staging/stage_tree.rs#L130
Do you see a solution?

Reply:
@malexandru-rdx plz see the latest commit with the `lazy_init!` that we discussed as an option

Reply:
Definitely better than having 1 extra field. I'm not sure about multithreading implications though (docs says it is behind an atomic lock for `Deref`).

Might need to change to `Option<>` at some point in the future anyway because of this. If nobody else feels strongly about it, I'd leave it like this for now.